src/ddmd/link.d

```diff
@@ -1039,4 +983,384 @@ version (Windows)
+                    return FileName.combine(VCToolsInstallDir, x64 ? r"lib\x64" : r"lib\x86");
+                return VCToolsInstallDir;
+            }
+            if (dir == VCDir.Bin)
+                return FileName.combine(VCInstallDir, x64 ? r"bin\amd64" : "bin");
+            if (dir == VCDir.Lib)
+                return FileName.combine(VCInstallDir, x64 ? r"lib\amd64" : "lib");
+            return VCInstallDir;
+        }
+
+        const(char)* getUCRTLibPath(bool x64)
+        {
+            if (UCRTSdkDir && UCRTVersion)
+               return PathBuilder(UCRTSdkDir) / "Lib" / UCRTVersion / (x64 ? r"ucrt\x64" : r"ucrt\x86");
+            return null;
+        }
+
+        const(char)* getSDKLibPath(bool x64)
+        {
+            if(WindowsSdkDir)
+            {
+                const(char)* arch = x64 ? "x64" : "x86";
+                auto sdk = PathBuilder(WindowsSdkDir) / "lib";
+                if(FileName.exists(sdk / WindowsSdkVersion / "um" / arch / "kernel32.lib")) // SDK 10.0
+                    return sdk / WindowsSdkVersion / "um" / arch;
+                else if(FileName.exists(sdk / r"win8\um" / arch / "kernel32.lib")) // SDK 8.0
+                    return sdk / r"win8\um" / arch;
+                else if(FileName.exists(sdk / r"winv6.3\um" / arch / "kernel32.lib")) // SDK 8.1
+                    return sdk / r"winv6.3\um" / arch;
+                else if(x64 && FileName.exists(sdk / arch / "kernel32.lib")) // SDK 7.1 or earlier
+                    return sdk / arch;
+                else if(!x64 && FileName.exists(sdk / "kernel32.lib")) // SDK 7.1 or earlier
+                    return sdk;
+            }
+            return null;
+        }
+
+        // iterate through subdirectories named by SDK version in baseDir and return the
+        //  one with the largest version that also contains the test file
+        static const(char)* findLatestSDKDir(const(char)* baseDir, const(char)* testfile)
+        {
+            auto allfiles = FileName.combine(baseDir, "*");
+            core.sys.windows.winbase.WIN32_FIND_DATAA fileinfo;
+            HANDLE h = FindFirstFileA(allfiles, &fileinfo);
+            if (h == INVALID_HANDLE_VALUE)
+                return null;
+
+            const(char)* res = null;
+            do
+            {
+                if (fileinfo.cFileName[0] >= '1' && fileinfo.cFileName[0] <= '9')
+                    if (res is null || strcmp(res, fileinfo.cFileName.ptr) < 0)
+                        if (FileName.exists(FileName.combine(FileName.combine(baseDir, fileinfo.cFileName.ptr), testfile)))
+                            res = strdup(fileinfo.cFileName.ptr);
+            }
+            while(FindNextFileA(h, &fileinfo));
+            FindClose(h);
+
+            return res;
+        }
+
+    }
+
+    struct PathBuilder
+    {
+        const(char)* p;
+        alias p this;
+
+        PathBuilder opDiv(const(char)* name)
+        {
+            if (!name)
```

Comment:
The uses I see of PathBuilder suggest that `name` is never null, so this struct is kinda pointless. In fact, `Filename.combine` should do the right thing with a `null` second argument anyway. (I checked and it does not, so that should be fixed there, not adding a wrapper here.)